### DISCUSSION TITLE
请问各位在飞书机器人里怎么发送二维码图片

### DISCUSSION THREAD

**Discussion by ycy0430** [tryzealot/zealot#1340] [General]
公司由钉钉转飞书了，钉钉机器人可以直接根据图片url发送图片，飞书不支持，要先将图片上传获取key才能发送二维码图片。zealot网络钩子里并不好直接实现。
难道要自己写一个服务接受zealot的图片，然后上传飞书，再发给飞书机器人吗

**Comment by ycy0430** [comment]
没注意文档里 有飞书 Interactive 卡片格式，我先去试试

**Reply by ycy0430** [reply]
发现还是不展示，只有图片的连接，就这样把，多点一下再扫

**Reply by icyleaf** [reply]
看来飞书改规则了，现在只有开发者和编辑器可以用完整版的 markdown 语法，精简版的 `lark_md` 不再支持外链图片地址，只能用上传的图片 key 或者用图片模块：https://open.feishu.cn/document/common-capabilities/message-card/message-cards-content/embedded-non-interactive-elements/image

**Reply by icyleaf** [reply]
文档已经移除图片展示的例子 😞

**Reply by mlch911** [reply]
直接用`interactive`类型，内容用`markdown`就行。然后还可以加上button类型，直接点button就可以安装了。

如果需要发图片，首先需要有图片文件（如果是url就先下载），然后用飞书的接口上传，获取到`image_key`，然后在`markdown`中就能使用，正常发图片，`url`用这个`image_key`就可以了。

**Comment by mlch911** [comment]
上个代码片段吧，`image_dic`是个`Hash`，`key`是原md中的图片url占位符，`value`是图片的url或者本地路径
飞书的api库用的是[lark-ruby-sdk](https://github.com/seandong/lark-ruby-sdk)
```ruby
# @param [String] token     机器人Token
# @param [String] title     消息标题
# @param [String] markdown  markdown内容。注意只支持部分 markdown 语法。
#                           若需要使用图片，markdown 中的图片链接为 image_dic 中的 key，value对应图片文件路径。会自动上传图片。
#                           详见https://open.feishu.cn/document/ukTMukTMukTM/uADOwUjLwgDM14CM4ATN。
# @param [Hash] btn_dic      配置按钮。key 为按钮文字，value 为按钮点击跳转的链接。
# @param [Hash] image_dic   上传图片。key 为 markdown 中的图片key，value 为图片文件路径。
def self.post_rich_content_to_feishu(token, title, markdown, btn_dic, image_dic)
  Lark.configure do |config|
    # 配置SDK
  end

  api = Lark::Api.new

  unless image_dic.empty?
    image_dic.each_pair do |key, image|
      uri = URI(image)
      image_name = nil
      if ['http', 'https'].include?(uri.scheme)
        image_name = "#{key}.png"
        `wget #{Shellwords.escape(uri.to_s)} -O #{image_name}`
        image = image_name
      end
      r = api.im.upload_image(image, 'message')
      `rm #{image_name}` if image_name
      markdown = markdown.sub(key.to_s, r.data['data']['image_key'])
    end
  end

  api.bot.send_markdown(token, title, markdown, btn_dic)
end
```

**Reply by ycy0430** [reply]
问题是没法先下载再上传图片，这功能需要作者支持一下

**Reply by mlch911** [reply]
我这就是先下载再上传啊😅

**Reply by ycy0430** [reply]
你是改作者源码的吗

**Reply by mlch911** [reply]
没有啊，你自己写个方法封装一下就行啊

写个发飞书的方法，参数有`markdown`和`image_dic`，然后复制我那段代码，就能下载图片然后再上传图片，之后再调[sdk](https://github.com/seandong/lark-ruby-sdk)里发消息的api就行

**Reply by ycy0430** [reply]
那还是要在项目里加代码， 这我一开始就想到了，所以想问问作者有没有其他办法的。

**Reply by mlch911** [reply]
最好用的是直接放点击安装的链接就行，不用扫码，也不用跳网页，扫码我都觉得麻烦😅

**Reply by ycy0430** [reply]
测试机上一般不登飞书，我们目前就是按钮和跳转连接都有，看他们需求。 话说最近测试反馈跳转到浏览器，二维码太小，背景还是黑色的，有点不太好扫上，还得手动放大二维码
